### PR TITLE
Add FingerprintJS library

### PR DESCRIPTION
**Pull request by chrisfromthelc**
Related to https://github.com/Automattic/gold/issues/460

## Proposed Changes

* This installs the FingerprintJS library for device fingerprinting on WordPress.com transactions.

## Why are these changes being made?

Project Thread: pbOQVh-4Zx-p2

This code will enable us to implement device fingerprinting for transactions on WordPress.com. These fingerprints are used as evidence during defense of chargebacks to card processors/banks.

## Testing Instructions

* No functional tests as it is just adding the library to the `packages` directory.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by pottedmeat** [comment]
I think you should add a reference to fingerprintjs from the parent tsconfig.json

**Review by pottedmeat — APPROVED** [review]
LGTM. I'd also encourage you to see if we missed this documentation somehow and/or get the [Monorepo document](https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md) updated based on what we figured out.

File: packages/fingerprintjs/tests/integration.test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import * as FPJS from '../src';
+
+/*
+ * Checks that the distributive file works in a browser.
+ * Build the minified script before running this test.
+ */
+
+declare const FingerprintJS: typeof FPJS;
+
+describe( 'Integration', () => {
```
Comment (tyxla) [inline packages/fingerprintjs/tests/integration.test.ts:10]: Note that these tests won't run right now. Tests run only for packages that have their own `jest.config.js` file.

File: packages/fingerprintjs/rollup.config.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+					file: 'resources/license_banner.txt',
+				},
+			},
+		} ),
+	],
+};
+
+const commonTerser = terserPlugin( terserConfig );
+
+const config: RollupOptions[] = [
```
Comment (tyxla) [inline packages/fingerprintjs/rollup.config.ts:38]: Ideally, the package would use Calypso's build system. Definitely not a requirement, and not work for the initial PR.

File: packages/fingerprintjs/karma.conf.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
+import { Config, CustomLauncher } from 'karma';
```
Comment (tyxla) [inline packages/fingerprintjs/karma.conf.ts:1]: Ideally, these tests would use Playwright, but this isn't a necessity right now, just a nice-to-have for consistency with the rest of the codebase.

File: packages/fingerprintjs/readme.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
+<p align="center">
```
Comment (tyxla) [inline packages/fingerprintjs/readme.md:1]: Should we add some information about where this came from and why it was forked? 

Also, it's not a big deal, but we usually capitalize the README.md file name.

File: packages/fingerprintjs/package.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		"rimraf": "^3.0.2",
+		"rollup": "^3.3.0",
+		"rollup-plugin-dts": "^5.2.0",
+		"rollup-plugin-license": "^3.0.1",
+		"terser-webpack-plugin": "^5.3.6",
+		"ts-loader": "^9.4.2",
+		"ts-node": "^10.9.1",
+		"typescript": "^4.9.3",
+		"ua-parser-js": "^0.7.32",
+		"webpack": "^5.75.0",
```
Comment (tyxla) [inline packages/fingerprintjs/package.json]: Ideally, we would use the same versions that we use across the repository. Not a blocker for this PR, but might start causing issues as we start integrating the package with the rest of Calypso.

File: packages/fingerprintjs/code_of_conduct.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
+# Contributor Covenant Code of Conduct
```
Comment (tyxla) [inline packages/fingerprintjs/code_of_conduct.md]: Calypso has its own COC (at the root of the repo); let's use that and remove this one.

File: packages/fingerprintjs/LICENSE [hunk 7]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2023 FingerprintJS, Inc
```
Comment (tyxla) [inline packages/fingerprintjs/LICENSE]: What kind of license is this? Should we include that it's GPL 2 or whatever it is? See the root LICENSE file.

File: packages/fingerprintjs/.eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
+module.exports = {
+	rules: {
+		'import/order': 'off', // Disable import order rule
+		'import/no-nodejs-modules': 'off', // Disable node modules import rule
```
Comment (tyxla) [inline packages/fingerprintjs/.eslintrc.js:4]: Ideally, we should follow similar or equal to the rest of the repo's coding standards. Not a blocker for the initial landing, obviously.

File: packages/fingerprintjs/resources/license_banner.txt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+FingerprintJS v<%= pkg.version %> - Copyright (c) FingerprintJS, Inc, <%= new Date().getFullYear() %> (https://fingerprint.com)
+Licensed under the MIT (http://www.opensource.org/licenses/mit-license.php) license.
```
Comment (tyxla) [inline packages/fingerprintjs/resources/license_banner.txt:2]: Ah, so there's where it says what the license is. Should we combine with the LICENSE file?

File: packages/fingerprintjs/src/agent.test.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
+import { version } from '../package.json';
+import { withMockProperties } from '../tests/utils';
+import { load as loadAgent } from './agent';
+import { sources } from './sources';
+import { hasScreenFrameBackup, resetScreenFrameWatch } from './sources/screen_frame';
+import { wait } from './utils/async';
+
+describe( 'Agent', () => {
```
Comment (tyxla) [inline packages/fingerprintjs/src/agent.test.ts:8]: These tests currently won't run. The package needs a `jest.config.js` file - see other packages. 

Consider this for the rest of the tests around here.

File: packages/fingerprintjs/karma.conf.ts (hunk 3 above)
Comment (tyxla) [inline packages/fingerprintjs/karma.conf.ts:1]: Note that you might need some custom work to make these tests actually run in CI - right now they won't run automatically.

File: packages/fingerprintjs/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,30 @@
+{
+	"extends": "@automattic/calypso-typescript-config/ts-package.json",
```
Comment (tyxla) [inline packages/fingerprintjs/tsconfig.json:2]: I appreciate we're already extending the default TS config, nice work 👏 

Let's make sure those files are actually being typechecked properly before landing.

**Review by tyxla — COMMENTED** [review]
Thanks for the ping 🙌 

I see no blockers for the initial merge. 

I've added some suggestions that can be addressed in follow-ups.

Note that none of the pre-existing tests will run, and some work will have to be done to run in CI and locally in the dev environment.

Feel free to ping me if you're struggling with any of it in the future. My team no longer focuses on Calypso itself, but we're happy to help and support anyone who still does.

File: packages/fingerprintjs/rollup.config.ts (hunk 2 above)
Comment (pottedmeat) [reply packages/fingerprintjs/rollup.config.ts:38]: I think we'll likely need to do this. I'm looking at the generated `dist` directory and the `package.json` points to `dist/fp.cjs.js` which I don't see.

File: packages/fingerprintjs/LICENSE (hunk 7 above)
Comment (chrisfromthelc) [reply packages/fingerprintjs/LICENSE]: I did some looking around, it seems like we should have the MIT license in place for that code, and a separate GPL license for our modifications.

I updated the main LICENSE to have the GPL text (from the Calypso license), and added a LICENSE-MIT file with text for that license (referenced from the main LICENSE file).

This seems like it should cover things, but let me know if this needs to change.

File: packages/fingerprintjs/resources/license_banner.txt (hunk 9 above)
Comment (chrisfromthelc) [reply packages/fingerprintjs/resources/license_banner.txt:2]: See https://github.com/Automattic/wp-calypso/pull/93475#discussion_r1726176715

**Comment by chrisfromthelc** [comment]
Merging to have the base code in production, and we are creating followup tasks for the test items.

**Pull request merged**